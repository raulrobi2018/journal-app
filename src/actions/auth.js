import {types} from "../types/types";

// En este caso este middleware devolverá un callback
//El dispatch lo obtiene de thunk
// En resumen esto es una función que dispara otra función luego que el setTimeout se resuelve
export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        setTimeout(() => {
            dispatch(login(123, "Raul"));
        }, 3500);
    };
};

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    };
};
