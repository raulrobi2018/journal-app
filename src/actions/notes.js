import { db } from "../firebase/firebase-config";
import { types } from '../types/types';

export const startNewNote = () => {
    // El segundo parámetro me da acceso al state actual, como lo hace el useSelector
    return async(dispatch, getState) => {
        // También se puede hacer desestructuración
        const uid = getState().auth.uid;

        const entry = {
            title: "",
            body: "",
            date: new Date().getTime()
        };

        // Grabo el registro en la base de datos
        const doc = await db.collection(`${uid}/journal/Notes`).add(entry);

        console.log(doc);

        dispatch(activeNote(doc.id, entry));
    };
};

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        // Todo lo que viene en la nota
        ...note
    }
});

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})