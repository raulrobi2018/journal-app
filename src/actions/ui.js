import {types} from "../types/types";

//Estas funciones hacen un return explicito.

export const setError = (err) => ({
    type: types.uiSetError,
    payload: err
});

export const removeError = () => ({
    type: types.uiRemoveError
});

export const startLoading = () => ({
    type: types.uiStartLoading
});

export const finishLoading = () => ({
    type: types.uiFinishLoading
});
