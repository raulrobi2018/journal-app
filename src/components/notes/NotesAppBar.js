import React from "react";

export const NotesAppBar = () => {
    return (
        <div className="notes_appbar">
            <span>25/02/2021</span>
            <div>
                <button className="btn">Picture</button>
                <button className="btn">Save</button>
            </div>
        </div>
    );
};
