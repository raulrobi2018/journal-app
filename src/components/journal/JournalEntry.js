import React from "react";
import moment from "moment";
import {activeNote} from "../../actions/notes";
import {useDispatch} from "react-redux";

export const JournalEntry = ({id, note}) => {
    const noteDate = moment(note.date);

    const dispatch = useDispatch();

    const handleActive = () => {
        dispatch(activeNote(id, note));
    };

    return (
        <div
            className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
            onClick={handleActive}
        >
            {note.url && (
                <div
                    className="journal__entry-picture"
                    style={{
                        backgroundSize: "cover",
                        backgroundImage: `url(${note.url})`
                    }}
                ></div>
            )}

            <div className="journal__entry-body">
                <p className="journal__entry-title">{note.title}</p>
                <p className="journal__entry-content">{note.body}</p>
            </div>

            <div className="journal__entry-date-box">
                <span>{noteDate.format("dddd")}</span>
                <h4>{noteDate.format("Do")}</h4>
            </div>
        </div>
    );
};
