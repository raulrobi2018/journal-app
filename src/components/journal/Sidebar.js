import React from "react";
import {JournalEntries} from "./JournalEntries";
import {useDispatch, useSelector} from "react-redux";
import {startLogout} from "../../actions/auth";
import {startNewNote} from "../../actions/notes";

export const Sidebar = () => {
    //Hook de Redux que retorna el state actual
    //En este caso tomo el atributo ui del state y luego desestructuro el mensaje
    const {name} = useSelector((state) => {
        return state.auth;
    });

    const dispatch = useDispatch();

    const handleLogout = () => {
        dispatch(startLogout());
    };

    const handleAddEntry = () => {
        dispatch(startNewNote());
    };

    return (
        <aside className="journal__sidebar">
            <div className="journal__sidebar-navbar">
                <h3 className="mt-5">
                    <i className="far fa-moon"></i>
                    <span> {name}</span>
                </h3>

                <button className="btn" onClick={handleLogout}>
                    Logout
                </button>
            </div>

            <div className="journal__new-entry" onClick={handleAddEntry}>
                <i className="far fa-calendar-plus fa-5x"></i>
                <p className="mt-5">New entry</p>
            </div>

            <JournalEntries />
        </aside>
    );
};
