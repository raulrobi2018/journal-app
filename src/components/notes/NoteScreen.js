import React, {useEffect, useRef} from "react";
import {useSelector} from "react-redux";
import {useForm} from "../../hooks/useForm";
import {NotesAppBar} from "./NotesAppBar";

export const NoteScreen = () => {
    const {active: note} = useSelector((state) => state.notes);
    const [formValues, {handleInputChange, reset}] = useForm(note);

    const {body, title} = formValues;

    // Cuando se seleccionan diferentes notas en pantalla, los campos no reflejan la informaci칩n
    // de la nota seleccionada, siempre se mantiene en la primera aunque la nota activa si se actualiza si miramos
    // por devtool de Redux, no se actualiza en nuestro useForm
    // Para solucionar esto se utiliza un useEffect y useRef y adem치s se modifica la funci칩n "reset" del useForm
    // Ver video 262 para los detalles de la explicaci칩n

    const activeId = useRef(note.id);

    useEffect(() => {
        if (note.id !== activeId.current) {
            reset(note);
            activeId.current = note.id;
        }
    }, [note, reset]);

    // ***********************************************************************

    return (
        <div className="notes__main-content">
            <NotesAppBar />

            <div className="notes__content">
                <input
                    type="text"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    autoComplete="off"
                    value={title}
                    onChange={handleInputChange}
                />

                <textarea
                    placeholder="What happened today"
                    className="notes__textarea"
                    value={body}
                    onChange={handleInputChange}
                ></textarea>

                {note.url && (
                    <div className="notes_image">
                        <img
                            src="https://imgd.aeplcdn.com/476x268/bw/ec/15278/Royal-Enfield-Classic-350-Side-87574.jpg?v=201711021421&q=80"
                            alt="Image"
                        />
                    </div>
                )}
            </div>
        </div>
    );
};
