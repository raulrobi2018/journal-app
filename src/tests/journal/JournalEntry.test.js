import React from "react";
import {mount, configure} from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router-dom";

//Configuración necesaria para que funcione mount con React 17
configure({adapter: new Adapter()});
import configureStore from "redux-mock-store"; //ES6 modules
import thunk from "redux-thunk";

import {JournalEntry} from "../../components/journal/JournalEntry";
import {activeNote} from "../../actions/notes";

// Configuracion necesaria para probar dispatch
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};

let store = mockStore(initState);

//Reemplaza la funcionalidad del dispatch del store por una función jest
store.dispatch = jest.fn();

const note = {
    id: 10,
    title: "Tremendo día!",
    body: "Hoy es feriado",
    date: 0,
    url: "https://tremendositio.com/dia.jpg"
};

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <JournalEntry id={note.id} note={note} />
        </MemoryRouter>
    </Provider>
);

describe("Testing JournalEntry component", () => {
    test("should display the component correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("should activate the note", () => {
        wrapper.find(".journal__entry").prop("onClick")();

        //Esta es otra forma de probar la llamada a una función de un componente
        //En este caso llamaría a handleActive
        expect(store.dispatch).toHaveBeenCalled();

        //Evalúo que se ejecute con los parámetros adecuados
        expect(store.dispatch).toHaveBeenCalledWith(activeNote(note.id, note));
    });
});
