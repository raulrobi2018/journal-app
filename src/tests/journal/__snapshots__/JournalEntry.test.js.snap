// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing JournalEntry component should display the component correctly 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://tremendositio.com/dia.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Tremendo d√≠a!
    </p>
    <p
      className="journal__entry-content"
    >
      Hoy es feriado
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
